packages/lib/services/e2ee/ppk/WebCryptoRsa.ts

```diff
@@ -0,0 +1,170 @@
+import { PublicKeyCrypto } from '../types';
+import type { webcrypto } from 'node:crypto';
+
+// The web crypto API is available in both NodeJS and browsers
+export type WebCryptoSlice = {
+	subtle: Pick<
+	SubtleCrypto,
+		// Restrict to methods that are known to be supported on mobile:
+		'generateKey'|'importKey'|'encrypt'|'decrypt'|'exportKey'
+	>;
+};
+
+interface KeyPair {
+	publicKey: CryptoKey;
+	privateKey: CryptoKey|null;
+}
+
+const isLowercaseHexadecimalString = (text: string) => {
+	return text.match(/^[a-f0-9]+$/) && text.length % 2 === 0;
+};
+
+const modulusLengthBits = 4096;
+export default class WebCryptoRsa implements PublicKeyCrypto<CryptoKeyPair> {
+	public constructor(private webCrypto_: WebCryptoSlice) {}
+
+	public static fromNodeCrypto(crypto: typeof webcrypto) {
+		return new WebCryptoRsa(
+			// Cast: Old versions of @types/node don't include crypto.subtle:
+			crypto as unknown as WebCryptoSlice,
+		);
+	}
+
+	public async generateKeyPair() {
+		// See the RSA dom example:
+		// https://github.com/mdn/dom-examples/blob/8c3d2a9781204e60b52371e300aebf067cd0b876/web-crypto/encrypt-decrypt/rsa-oaep.js#L71
+		// and the relevant MDN documentation:
+		// https://developer.mozilla.org/en-US/docs/Web/API/RsaHashedKeyGenParams
+		const keyPair = await this.webCrypto_.subtle.generateKey({
+			name: 'RSA-OAEP',
+			modulusLength: modulusLengthBits,
+			// From https://developer.mozilla.org/en-US/docs/Web/API/RsaHashedKeyGenParams#publicexponent
+			// "Unless you have a good reason to use something else, specify 65537 here ([0x01, 0x00, 0x01])."
+			publicExponent: new Uint8Array([1, 0, 1]),
+			hash: 'SHA-256',
+		}, true, ['encrypt', 'decrypt']);
+
+		return {
+			keyPair,
+			keySize: modulusLengthBits,
+		};
+	}
+
+	public async loadKeys(publicKeySource: string, privateKeySource: string): Promise<KeyPair> {
+		const importKey = (keySource: string, usages: KeyUsage[]) => {
+			return this.webCrypto_.subtle.importKey(
+				'jwk', JSON.parse(keySource), { name: 'RSA-OAEP', hash: 'SHA-256' }, true, usages,
+			);
+		};
+		const publicKey = await importKey(publicKeySource, ['encrypt']);
+		const privateKey = privateKeySource ? await importKey(privateKeySource, ['decrypt']) : null;
+		return {
+			publicKey,
+			privateKey,
+		};
+	}
+
+	private textToBuffer_(textUtf8: string) {
+		// To avoid data loss when restoring (whether everything is capital or lowercase), work only
+		// with lowercase hexadecimal.
+		const isHex = isLowercaseHexadecimalString(textUtf8);
```

Comment:
I'm not sure I understand the logic here - is that for compatibility with the previous version? Which means some data is hex and some is not?

Reply:
Currently, Joplin seems to only provide non-hex-encoded data during tests (assuming that Joplin master keys are always hex-encoded).

An alternative to this might be to:
- `throw` when provided non-hex-encoded data, both here and in the legacy RSA implementation. This would help ensure that the input data remains hexadecimal encoded in the future. 
- Add tests and/or runtime assertions elsewhere to help ensure that Joplin master keys remain hex-encoded in the future (even if the legacy RSA implementation is still in-use).
   - Without hexadecimal decoding, Joplin master keys are larger than the maximum plaintext size (see notes above).